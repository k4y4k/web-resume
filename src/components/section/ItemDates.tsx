import * as React from 'react'
import tw, { TwStyle } from 'twin.macro'
import customParseFormat from 'dayjs/plugin/customParseFormat'
import dayjs from 'dayjs'
import timezone from 'dayjs/plugin/timezone'
import utc from 'dayjs/plugin/utc'

interface ItemDatesTypes {
  to?: string
  from: string
}

const datesStyles = (): Array<TwStyle | string> => [
  tw`inline-block`,
  tw`print:text-sm`,
]

const ItemDates = ({ to, from }: ItemDatesTypes): JSX.Element | null => {
  if (from === '') return null

  dayjs.extend(customParseFormat)
  dayjs.extend(utc)
  dayjs.extend(timezone)

  const toDate = dayjs(to, ['YYYY-MM-DD', 'YYYY-MM', 'YYYY'], true)
  const fromDate = dayjs(from, ['YYYY-MM-DD', 'YYYY-MM', 'YYYY'], true)

  const isFromDateValid = fromDate.isValid()
  const isToDateValid = toDate.isValid()

  // no to date / "to present"
  if (from !== '' && to === '')
    return (
      <p css={datesStyles()} data-testid='sectionItemDates'>
        {fromDate.format('MMM YYYY')} - Present
      </p>
    )

  // dates not valid / supplied to date invalid
  if (!isFromDateValid || !isToDateValid)
    return (
      <p data-testid='sectionItemDates' css={datesStyles()}>
        Error: Dates Not Valid
      </p>
    )

  // to later than from
  if (fromDate > toDate)
    return (
      <p data-testid='sectionItemDates' css={datesStyles()}>
        Error: Dates Not Chronological
      </p>
    )

  // just years
  // we assume that anything set to Jan 01 20XX just wants 20XX

  // // get 20XX
  const fromDateYear = fromDate.get('year')
  const toDateYear = toDate.get('year')

  // check for sameness
  if (
    fromDate.isSame(`${fromDateYear}-01-01`, 'day') &&
    toDate.isSame(`${toDateYear}-01-01`, 'day')
  )
    if (fromDate.isSame(toDate, 'year'))
      // have two 20XX "just years" dates. Now, are they the same year?
      return (
        // yes
        <p css={datesStyles()} data-testid='sectionItemDates'>
          {fromDate.format('YYYY')}
        </p>
      )
    else
      return (
        // no
        <p css={datesStyles()} data-testid='sectionItemDates'>
          {fromDate.format('YYYY')} - {toDate.format('YYYY')}
        </p>
      )

  // same date (one day event?)
  if (fromDate.isSame(toDate, 'day'))
    return (
      <p css={datesStyles()} data-testid='sectionItemDates'>
        {toDate.format('MMM YYYY')}
      </p>
    )

  return (
    <p css={datesStyles()} data-testid='sectionItemDates'>
      {fromDate.format('MMM YYYY')} - {toDate.format('MMM YYYY')}
    </p>
  )
}

export default ItemDates
