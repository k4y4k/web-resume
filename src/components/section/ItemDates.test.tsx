import * as React from 'react'
import { screen, render } from '@testing-library/react'
import ItemDates from './ItemDates'

describe('<ItemDates />', () => {
  test('handles no dates', () => {
    render(<ItemDates from='' />)

    const sectionItemDates = screen.getByText('Error: No Item Dates')
    expect(sectionItemDates).toBeInTheDocument()
  })

  describe('handles badly formatted dates', () => {
    render(<ItemDates from='ABCD8' to='2020' />)
    const sectionItemDates = screen.getByText('Error: Dates Not Valid')

    expect(sectionItemDates).toBeInTheDocument()
  })

  test('bad from date', () => {
    render(<ItemDates from='2090' to='2007-05' />)
    const sectionItemDates = screen.getByText('Error: Dates Not Valid')

    expect(sectionItemDates).toBeInTheDocument()
  })

  test('bad to and from date', () => {
    render(<ItemDates to='111111' from='29470' />)

    const sectionItemDates = screen.getByText('Error: Dates Not Valid')
    expect(sectionItemDates).toBeInTheDocument()
  })
})

test('start date -> end date', () => {
  render(<ItemDates from='1970-01-01' to='2021-01-28' />)

  const sectionItemDates = screen.getByTestId('sectionItemDates')
  expect(sectionItemDates).toHaveTextContent('Jan 1970 - Jan 2021')
})

test('start date -> present (no end date)', () => {
  render(<ItemDates from='2012-05' />)

  const sectionItemDates = screen.getByTestId('sectionItemDates')
  expect(sectionItemDates).toBeInTheDocument()
  expect(sectionItemDates).toHaveTextContent('May 2012 - Present')
})

test('day, month, year date expressed as Month, Year', () => {
  render(<ItemDates from='1991-12-25' />)

  const sectionItemDates = screen.getByTestId('sectionItemDates')
  expect(sectionItemDates).toBeInTheDocument()
  expect(sectionItemDates).toHaveTextContent('Dec 1991 - Present')
})

test('month, year date expressed as Month, Year', () => {
  render(<ItemDates from='2020-10-27' to='2021-07-19' />)

  const sectionItemDates = screen.getByTestId('sectionItemDates')
  expect(sectionItemDates).toBeInTheDocument()
  expect(sectionItemDates).toHaveTextContent('Oct 2020 - Jul 2021')
})

test('dates must make chronological sense', () => {
  render(<ItemDates from='2021-06-14' to='2020-04-22' />)

  const sectionItemDates = screen.getByText('Error: Dates Not Chronological')
  expect(sectionItemDates).toBeInTheDocument()
})
