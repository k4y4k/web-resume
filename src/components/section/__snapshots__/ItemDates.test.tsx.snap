// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemDates /> dates must make chronological sense 1`] = `
<p
  data-testid="sectionItemDates"
>
  Error: Dates Not Chronological
</p>
`;

exports[`<ItemDates /> handles badly formatted dates 1`] = `
<p
  data-testid="sectionItemDates"
>
  Error: Dates Not Valid
</p>
`;

exports[`<ItemDates /> handles no dates 1`] = `null`;

exports[`<ItemDates /> month, year date expressed as Month, Year 1`] = `
<p
  data-testid="sectionItemDates"
>
  Oct 2020
   - 
  Jul 2021
</p>
`;

exports[`<ItemDates /> same dates: just one year 1`] = `
<p
  data-testid="sectionItemDates"
>
  1999
</p>
`;

exports[`<ItemDates /> same dates: years 1`] = `
<p
  data-testid="sectionItemDates"
>
  2012
   - 
  2013
</p>
`;

exports[`<ItemDates /> start date -> end date 1`] = `
<p
  data-testid="sectionItemDates"
>
  Jan 1970
   - 
  Jan 2021
</p>
`;

exports[`<ItemDates /> start date -> present (no end date) 1`] = `
<p
  data-testid="sectionItemDates"
>
  May 2012
   - Present
</p>
`;
